test/unit/uploader.basic.api.js

```diff
@@ -505,4 +505,32 @@ describe("uploader.basic.api.js", function () {
             assert.equal(formattedSize, "9.6GB");
         });
     });
+
+    describe("_validateFileOrBlobData", function() {
+        var originalFileOrInput = qq.isFileOrInput;
+        beforeEach(function () {
+            fineuploader = new qq.FineUploaderBasic();
+        });
+        afterEach(function() {
+            qq.isFileOrInput = originalFileOrInput;
+        });
+
+        it("fails if file is empty and allowEmpty is false", function() {
+            qq.isFileOrInput = function() { return true; };
+            fineuploader._fileOrBlobRejected = function() {};
+            var validationDescriptor = { size: 0 };
+
+            fineuploader._validateFileOrBlobData({}, validationDescriptor)
+                .then(function() { assert.fail(); }, function() {});
```

Comment:
This (and the subsequent test) do not look valid. As far as I can tell, the test will exit before the promise is resolved. This is effectively an async test. You'll need to make use of the `done` function in both tests.

In the first test, in the second function (failure), add a call to `done`. In the second test, add a call to `done` in the first `then` function (success).